test/bigquery.ts

```diff
@@ -708,26 +735,207 @@ describe('BigQuery', () => {
       );
 
       assert.throws(() => {
-        BigQuery.getType_(undefined);
+        BigQuery.getTypeDescriptorFromValue_(undefined);
+      }, expectedError);
+    });
+
+    it('should throw with an empty array', () => {
+      const expectedError = new RegExp(
+        'Type must be provided for empty array.'
+      );
+
+      assert.throws(() => {
+        BigQuery.getTypeDescriptorFromValue_([]);
       }, expectedError);
     });
+
+    it('should throw with an null value', () => {
+      const expectedError = new RegExp(
+        'Type must be provided for null values.'
+      );
+
+      assert.throws(() => {
+        BigQuery.getTypeDescriptorFromValue_(null);
+      }, expectedError);
+    });
+  });
+
+  describe('getTypeDescriptorFromProvidedType_', () => {
+    const QUERY_STRING = 'SELECT * FROM [dataset.table]';
+    const NAMED_PARAMS = {
+      key: 'value',
+    };
+    const POSITIONAL_PARAMS = ['value'];
+
+    const NAMED_TYPES = {key: 'STRING'};
+
+    const POSITIONAL_TYPES = ['STRING'];
+
+    it('should return correct type for an array', () => {
+      const type = BigQuery.getTypeDescriptorFromProvidedType_(['INT64']);
+
+      assert.deepStrictEqual(type, {
+        type: 'ARRAY',
+        arrayType: {
+          type: 'INT64',
+        },
+      });
+    });
+
+    it('should return correct type for a struct', () => {
+      const type = BigQuery.getTypeDescriptorFromProvidedType_({prop: 'INT64'});
+
+      assert.deepStrictEqual(type, {
+        type: 'STRUCT',
+        structTypes: [
+          {
+            name: 'prop',
+            type: {
+              type: 'INT64',
+            },
+          },
+        ],
+      });
+    });
+
+    it('should throw for invalid provided types with named params', done => {
+      const INVALID_TYPES = {key: 'invalid'};
+
+      assert.throws(() => {
+        bq.createQueryJob({
+          query: QUERY_STRING,
+          params: NAMED_PARAMS,
+          types: INVALID_TYPES,
+        });
+      }, /Invalid type provided./);
+      done();
```

Comment:
This done() and the ones that follow in this block can be removed. It's only needed when we want to hold the test from exiting prematurely while we wait for a callback to be executed. In this case, the `assert.throws()` will run sync and then exit the test block with a failure or a success.

Reply:
I'm really curious about this because these tests that check for errors keep timing out if I don't call done(). Is there something else I should be doing to fix that?

Reply:
I think you just need to remove the `done` argument from the test function's argument list:

```diff
- it('should ...', done => {
+ it('should ...', () => {
```

Reply:
Oh, got it.